### PR TITLE
refactor(*): typescript rewrite

### PR DESCRIPTION
**Pull request by Mesteery**

### CODE REVIEW

File: src/core/WebmBase.ts [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+  /**
+   * Creates a new Webm demuxer.
+   * @param [options] options that you would pass to a regular Transform stream.
+   */
+  public constructor(options: TransformOptions = {}) {
+    super({ readableObjectMode: true, ...options });
+  }
+
+  public _transform(chunk: Buffer, encoding: BufferEncoding, done: TransformCallback): void {
```
Comment (B-Derouet) [inline src/core/WebmBase.ts:58]: private ?

File: src/core/WebmBase.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+  /**
+   * Takes a buffer and attempts to read and process a tag.
+   * @private
+   * @param chunk the buffer to read from.
+   * @param offset the offset in the buffer.
+   * @returns contains the new `offset` (number) and optionally the `_skipUntil` property,
+   * indicating that the stream should ignore any data until a certain length is reached.
+   * Returns the WebmBaseDemuxer.TOO_SHORT symbol if the data wasn't big enough to facilitate the request.
+   */
+  public _readTag(chunk: Buffer, offset: number): Tag | typeof WebmBaseDemuxer.TOO_SHORT {
```
Comment (B-Derouet) [inline src/core/WebmBase.ts:145]: private ?

File: src/core/WebmBase.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    }
+    return { offset: offset + dataLength };
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  protected _checkHead(data: Buffer): void {
+    throw new Error('Method not implemented.');
+  }
+
+  public _destroy(error: Error | null, callback: (error: Error | null) => void): void {
```
Comment (B-Derouet) [inline src/core/WebmBase.ts:214]: private ?

File: src/core/WebmBase.ts [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    this.cleanup();
+    callback(error);
+  }
+
+  public _final(callback: () => void): void {
```
Comment (B-Derouet) [inline src/core/WebmBase.ts:219]: private ?

File: src/core/WebmBase.ts [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+  /**
+   * Cleans up the demuxer when it is no longer required.
+   */
+  private cleanup() {
+    this._remainder = undefined;
+    this._incompleteTrack = {};
+  }
+}
+
+function vintLength(buffer: Buffer, index: number): number | typeof WebmBaseDemuxer.TOO_SHORT {
```
Comment (B-Derouet) [inline src/core/WebmBase.ts:233]: Il serait plus intéressant de bouger ces fonctions dans des fichiers à part

File: src/opus/OggDemuxer.ts [hunk 6]
```diff
@@ -9,45 +9,43 @@ const OPUS_TAGS = Buffer.from('OpusTag');
...
-  constructor(options = {}) {
+  public constructor(options: TransformOptions = {}) {
     super({ readableObjectMode: true, ...options });
-    this._remainder = undefined;
-    this._head = undefined;
-    this._bitstream = undefined;
   }
 
-  _transform(chunk, encoding, done) {
+  public _transform(chunk: Buffer, encoding: BufferEncoding, done: TransformCallback): void {
```
Comment (B-Derouet) [inline src/opus/OggDemuxer.ts:25]: private ?

File: src/opus/OggDemuxer.ts [hunk 7]
```diff
@@ -100,39 +109,22 @@ class OggDemuxer extends Transform {
     return chunk.slice(start);
   }
 
-  _destroy(err, cb) {
-    this._cleanup();
-    return cb ? cb(err) : undefined;
+  public _destroy(error: Error | null, callback: (error: Error | null) => void): void {
```
Comment (B-Derouet) [inline src/opus/OggDemuxer.ts:112]: private ?

File: src/opus/OggDemuxer.ts [hunk 8]
```diff
@@ -100,39 +109,22 @@ class OggDemuxer extends Transform {
...
+    this.cleanup();
+    callback(error);
   }
 
-  _final(cb) {
-    this._cleanup();
-    cb();
+  public _final(callback: () => void): void {
```
Comment (B-Derouet) [inline src/opus/OggDemuxer.ts:117]: private ?

**Review by B-Derouet — CHANGES_REQUESTED** [review]

File: test/util.ts [hunk 9]
```diff
@@ -5,19 +5,22 @@ export function roughlyEquals(x: Buffer, y: Buffer): boolean {
...
   return true;
 }
 
 export function streamToBuffer(stream: Stream): Promise<Buffer> {
   return new Promise((resolve, reject) => {
     const chunks: Buffer[] = [];
+
     stream.on('data', (chunk: Buffer) => chunks.push(chunk));
-    stream.on('error', reject);
+    stream.on('error', () => reject());
```
Comment (Mesteery) [inline test/util.ts:23]: Why ?

Comment (B-Derouet) [reply test/util.ts:23]: Oh, I forgot to revert this one...

Comment (B-Derouet) [reply test/util.ts:23]: But it's not very important, works the same :+1:

Comment (Mesteery) [reply test/util.ts:23]: Yes, and not.
Before, the Error is rejected.
With your changes, the Error is not rejected.

**Review by B-Derouet — APPROVED** [review]
LGTM

**Pull request merged**